pkg/diagnostics/grpc_tracing.go

```diff
@@ -98,73 +101,90 @@ func GRPCTraceUnaryServerInterceptor(appID string, spec config.TracingSpec) grpc
 }
 
 // GRPCTraceStreamServerInterceptor sets the trace context or starts the trace client span based on request.
+// This is used by proxy requests too.
 func GRPCTraceStreamServerInterceptor(appID string, spec config.TracingSpec) grpc.StreamServerInterceptor {
-	return func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
-		if strings.Index(info.FullMethod, daprPackagePrefix) == 0 {
-			return handler(srv, ss)
-		}
-
-		var span trace.Span
-		spanName := info.FullMethod
+	return func(srv any, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		var (
+			span             trace.Span
+			spanKind         trace.SpanStartOption
+			prefixedMetadata map[string]string
+			reqSpanAttr      map[string]string
+			isProxied        bool
+		)
 
 		ctx := ss.Context()
-		md, _ := metadata.FromIncomingContext(ctx)
-
-		vals := md[GRPCProxyAppIDKey]
-		if len(vals) == 0 {
-			return errors.Errorf("cannot proxy request: missing %s metadata", GRPCProxyAppIDKey)
-		}
-
-		targetID := vals[0]
-		wrapped := grpcMiddleware.WrapServerStream(ss)
-		sc, _ := SpanContextFromIncomingGRPCMetadata(ctx)
 
-		var spanKind trace.SpanStartOption
-
-		if appID == targetID {
+		// This middleware is shared by multiple services and proxied requests, which need to be handled separately
+		switch {
+		// For gRPC service invocation, this generates ServerSpan
+		case strings.HasPrefix(info.FullMethod, daprInternalPrefix):
 			spanKind = trace.WithSpanKind(trace.SpanKindServer)
-		} else {
+
+		// For gRPC API, this generates ClientSpan
+		case strings.HasPrefix(info.FullMethod, daprRuntimePrefix):
 			spanKind = trace.WithSpanKind(trace.SpanKindClient)
+
+		// For proxied requests, this generates a span depending on whether this is the server (target) or client
+		default:
+			isProxied = true
```

Comment:
Just wanted to understand one thing here: any calls from "/dapr.proto.operator." etc. will fall in this category, so based on GRPCProxyAppIDKey, it would be decided if Server span to be allocated or client span to be allocated. Ideally, GRPCProxyAppIDKey would be present here, as it is dapr-app-id. Just wondering, would calls like these from control plane service etc. be falling under server or client category? 
Or there is a gap in my understanding here.

Reply:
I may be wrong but I don't believe we have instrumented calls from the runtime to the operator service - I don't think those calls would ever go through this.